base_layer/core/src/base_node/service/service.rs

```diff
@@ -280,222 +309,223 @@ where B: BlockchainBackend + 'static
+                    "Failed to send outbound request from Base Node: {}", resp
+                );
+                Err(resp)
+            }));
         }
-
-        Ok(())
     }
 
-    async fn handle_outbound_request(
-        &mut self,
-        reply_tx: OneshotSender<Result<Vec<NodeCommsResponse>, CommsInterfaceError>>,
-        request: NodeCommsRequest,
-        request_type: NodeCommsRequestType,
-    ) -> Result<(), CommsInterfaceError>
-    {
-        let request_key = generate_request_key(&mut OsRng);
-        let service_request = proto::BaseNodeServiceRequest {
-            request_key,
-            request: Some(request.into()),
-        };
-
-        let mut send_msg_params = SendMessageParams::new();
-
-        match request_type {
-            NodeCommsRequestType::Single => send_msg_params.random(1),
-            NodeCommsRequestType::Many => send_msg_params.neighbours(Vec::new()),
-        };
-
-        let send_result = self
-            .outbound_message_service
-            .send_message(
-                send_msg_params.finish(),
-                OutboundDomainMessage::new(TariMessageType::BaseNodeRequest, service_request),
-            )
-            .await
-            .map_err(|e| CommsInterfaceError::OutboundMessageService(e.to_string()))?;
-
-        match send_result.resolve_ok().await {
-            Some(tags) if tags.is_empty() => {
-                let _ = reply_tx
-                    .send(Err(CommsInterfaceError::NoBootstrapNodesConfigured))
-                    .or_else(|resp| {
-                        error!(
-                            target: LOG_TARGET,
-                            "Failed to send outbound request from Base Node as no bootstrap nodes were configured"
-                        );
-                        Err(resp)
-                    });
-            },
-            Some(tags) => {
-                let dest_count = tags.len();
-                // Wait for matching responses to arrive
-                self.waiting_requests.insert(request_key, WaitingRequest {
+    Ok(())
+}
+
+async fn handle_outbound_request(
+    executor: runtime::Handle,
+    mut outbound_message_service: OutboundMessageRequester,
+    waiting_requests: WaitingRequests,
+    timeout_sender: Sender<RequestKey>,
+    reply_tx: OneshotSender<Result<Vec<NodeCommsResponse>, CommsInterfaceError>>,
+    request: NodeCommsRequest,
+    request_type: NodeCommsRequestType,
+    config: BaseNodeServiceConfig,
+) -> Result<(), CommsInterfaceError>
+{
+    let request_key = generate_request_key(&mut OsRng);
+    let service_request = proto::BaseNodeServiceRequest {
+        request_key,
+        request: Some(request.into()),
+    };
+
+    let mut send_msg_params = SendMessageParams::new();
+
+    match request_type {
+        NodeCommsRequestType::Single => send_msg_params.random(1),
+        NodeCommsRequestType::Many => send_msg_params.neighbours(Vec::new()),
+    };
+
+    let send_result = outbound_message_service
```

Comment:
This should only be sent after the request_key has been added to WaitingRequest. If it fails, then remove it from WaitingRequest on the way out

Reply:
The order can't really be changed as the outbound message service first needs to provide how many peers will the message be sent to before the waiting request can be created. This behaviour will probably change in the near future.